File: core/core/image.go
package core

import (
	"github.com/Tracer1337/gg"
)

func (e *ImageElement) GetIndex() int {
	return e.Index
}

func (e *ImageElement) GetType() string {
	return "image"
}

func (e *ImageElement) Draw(dc *gg.Context, c *Canvas, i int) {
	defer dc.Identity()
	defer dc.ResetClip()

	e.Rect.ApplyRotation(dc)
	e.drawBorderRadius(dc, c)
	e.drawImage(dc, c)

	if c.Debug {
		e.Rect.Draw(dc)
	}
}

func (e *ImageElement) drawBorderRadius(dc *gg.Context, c *Canvas) {
	dc.DrawRoundedRectangle(e.Rect.X, e.Rect.Y, e.Rect.Width, e.Rect.Height, e.Data.BorderRadius)
	dc.Clip()
}

func (e *ImageElement) drawImage(dc *gg.Context, c *Canvas) {
	bounds := e.Data.Image.Bounds()
	sx, sy := e.Rect.ApplyScaling(dc, float64(bounds.Max.X), float64(bounds.Max.Y))
	dc.DrawImage(e.Data.Image, int(e.Rect.X*(1/sx)), int(e.Rect.Y*(1/sy)))
	dc.Scale(1/sx, 1/sy)
}

File: concept/src/image.ts
import { loadImage } from "canvas"
import { DrawingContext } from "./draw"
import { RenderingContext } from "./render"
import { ImageElement as IImageElement, ImageData, Rect } from "./struct"

export class ImageElement implements IImageElement {
    public index: number
    public rect: Rect
    public data: ImageData
    
    public getIndex() {
        return this.index
    }

    public getType() {
        return "image"
    }

    public async draw(rc: RenderingContext, dc: DrawingContext, i: number) {
        this.rect.applyRotation(dc)
        this.drawBorderRadius(rc, dc)
        await this.drawImage(rc, dc)

        if (rc.canvas.debug) {
            this.rect.draw(dc)
        }
    }

    private drawBorderRadius(rc: RenderingContext, dc: DrawingContext) {

    }

    private async drawImage(rc: RenderingContext, dc: DrawingContext) {
        const image = await loadImage(this.data.image)
        dc.cc.drawImage(image, this.rect.x, this.rect.x, this.rect.width, this.rect.height)
    }
}
